### DISCUSSION TITLE
tcp client/server not support for macos?

### DISCUSSION THREAD

**Discussion by xixioba** [wasmerio/wasmer#3770] [Q&A] [answered]
```
cargo run --features compiler,cranelift,debug -- --enable-threads ../wasi/tests/example-tcp-client.wasm
```
return 
```
Failed to connect: Not supported (os error 58)
Finished.
```
Can I be of some help?

**Comment by ptitSeb** [comment, marked as answer]
you need to use `--net` to enable network. Use `--help` for more details.

**Reply by xixioba** [reply]
Thanks, I just followed `lib/wasi/README.md` and ignored `--help`

**Comment by xixioba** [comment]
@ptitSeb Also, I'm interested in `experimental-io-devices`, but I can't find a way to get it working properly, any instructions that work?

**Reply by ptitSeb** [reply]
I don't use that, so I'm not sure. But you should open another ticket, it's a really different question there

**Reply by xixioba** [reply]
Ok, I'm going to open another ticket.

**Comment by xixioba** [comment]
@ptitSeb Finally, as a noob question, if I write wasm with tcp/udp function, can I embed it in a non-rust environment? (e.g. c++ or python?)🤔

**Reply by ptitSeb** [reply]
Yes, there is a C-API so you can use WebAssembly everywhere. 
You should here: https://www.reddit.com/r/rust/comments/120o4yd/build_universal_libraries_with_rust/ for some help on how to do it.

**Reply by xixioba** [reply]
Thanks!
